lib/rubocop/cop/rspec/file_path.rb

```diff
@@ -102,8 +149,30 @@ def ignore_methods?
           cop_config['IgnoreMethods']
         end
 
+        def file_extensions
+          cop_config.fetch('SpecFileExtensions', ['_spec.rb'])
+        end
+
+        def glob_metachars
+          ['*', '?', '{', '}', '[', ']', ',']
+        end
+
+        def escape_glob(str)
+          (str.split.uniq & glob_metachars)
+            .reduce(str) { |prior, mc| prior.gsub(mc, "\\#{mc}") }
+        end
+
+        def extension_glob
+          if file_extensions.count == 1
+            '*' + escape_glob(file_extensions.first)
+          else
+            "*{#{file_extensions.map(&method(:escape_glob)).join(',')}}"
+          end
+        end
+
         def filename_ends_with?(glob)
-          File.fnmatch?("*#{glob}", processed_source.buffer.name)
+          flags = file_extensions.count > 1 ? [File::FNM_EXTGLOB] : []
+          File.fnmatch?("*#{glob}", processed_source.buffer.name, *flags)
```

Comment:
There's [a recent pull request](https://github.com/rubocop-hq/rubocop-rspec/pull/869/files#diff-7b5894acb0fe1766bf72709ffae459a6R106), not sure which one gets merged first.
Do you think changes will be easy to merge?

Reply:
TL, DR: I think both your suggestions are good.  If you like the overall idea and think it might be generally useful, I'll improve it in accordance with your suggestions and either re-awaken this pull request or submit another one.
---

The existing configuration doesn't quite solve the problem. Setting the pattern in the RSpec config tells RSpec which files contain specs. Setting AllCops/RSpec/Patterns tells Rubocop::RSpec which Ruby files are more properly RSpec files. But when you get to the RSpec/FilePath cop, the only acceptable extension is a hardcoded `_spec.rb`. In fact, until you've set the RSpec config pattern and the Rubocop::RSpec config pattern to something that doesn't match `_spec.rb`, there's nothing about the extension to upset RSpec/FilePath.

I originally considered this as a band-aid patch for something that irked me. The original logic was to take the class structure of the code under test and use it to construct a glob that should match the file path of the spec file. So I stayed consistent with that logic, but altered the construction of the glob to match multiple extensions.

The question RSpec/FilePath tries to answer is this: given the class structure of the code under test, are the pathname and filename of the spec file consistent with Ruby's and RSpec's conventions?  In this context, I think it's best separated into two questions: are the pathname and filename consistent with the classes tested in the spec file?  and is the extension on the spec file one that RSpec, Rubocop::RSpec, and programmers will recognize as a spec file?

Dividing this into two rubrics is probably a better solution than doing elaborate gymnastics with strings and escaping and glob patterns to create a more complex (unmaintainable mess of a) glob pattern (that looks like bad Perl I wrote in my younger days) to answer both questions at the same time.  Pulling the acceptable file extensions from the RSpec and Rubocop config files is just a bonus.

And with all that said: since the best approach is likely to revert to the old behavior, but removing extensions altogether from the glob pattern, and add a second test for just the extension, the two pull requests are not likely to conflict. They're not in conflict logically: my change says "if there are two or more acceptable filename extensions, we need to pass `File::FNM_EXTGLOB` as a flag to `File.fnmatch?`" while the other pull request says "we need to handle the case when the path might be weird because we invoked rubocop from somewhere other than the project root, so we need to munge it before we pass it to `File.fnmatch?`. This is the sort of thing that casues easy merge conflicts because they both operate in fiddly ways on the same line of code, but they're easy for a human to resolve.

So.  I will revise the patch to make it better, and we can consider the new improved version  in a few days.

Reply:
It's hard to disagree with your point that this cop does the job of several cops. In addition to matching file name and making sure extension is proper, I can think of additional duties that it could do, but that would really make it a god cop:
 - make sure files containing root `shared_examples` and `shared_context` have the right path names
 - make sure they have a proper extension that **doesn't** match RSpec's pattern (also reflected in `AllCops/RSpec/Pattern` configuration item.

Any kind of improvement on this front is appreciated, if you could extract the part checking just the extension from `FilePath` without compromising its current responsibilities, that would be just great.